platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/DefaultRootTestEventReporter.java

```diff
@@ -17,26 +17,68 @@
 package org.gradle.api.internal.tasks.testing;
 
 import org.gradle.api.NonNullApi;
+import org.gradle.api.internal.tasks.testing.results.HtmlTestReportGenerator;
+import org.gradle.api.internal.tasks.testing.results.TestExecutionResultsListener;
 import org.gradle.api.internal.tasks.testing.results.TestListenerInternal;
 import org.gradle.api.tasks.VerificationException;
 import org.gradle.internal.id.IdGenerator;
+import org.gradle.internal.logging.ConsoleRenderer;
 import org.gradle.util.internal.TextUtil;
 
+import java.nio.file.Path;
 import java.time.Instant;
 
 @NonNullApi
 class DefaultRootTestEventReporter extends DefaultGroupTestEventReporter {
+
+    private final Path testReportDirectory;
+    private final RootTestEventRecorder binaryResultsRecorder;
+    private final HtmlTestReportGenerator htmlTestReportGenerator;
+    private final TestExecutionResultsListener executionResultsListener;
+
+    // Mutable state
     private String failureMessage;
 
-    DefaultRootTestEventReporter(TestListenerInternal listener, IdGenerator<?> idGenerator, TestDescriptorInternal testDescriptor) {
-        super(listener, idGenerator, testDescriptor, new TestResultState(null));
+    DefaultRootTestEventReporter(
+        String rootName,
+        TestListenerInternal listener,
+        IdGenerator<?> idGenerator,
+        Path testReportDirectory,
+        RootTestEventRecorder binaryResultsRecorder,
+        HtmlTestReportGenerator htmlTestReportGenerator,
+        TestExecutionResultsListener executionResultsListener
+    ) {
+        super(
+            listener,
+            idGenerator,
+            new DefaultTestSuiteDescriptor(idGenerator.generateId(), rootName),
+            new TestResultState(null)
+        );
+
+        this.testReportDirectory = testReportDirectory;
+        this.binaryResultsRecorder = binaryResultsRecorder;
+        this.htmlTestReportGenerator = htmlTestReportGenerator;
+        this.executionResultsListener = executionResultsListener;
     }
 
     @Override
     public void close() {
         super.close();
-        if (failureMessage != null) {
-            throw new VerificationException(failureMessage);
+
+        // Ensure binary results are written to disk.
+        binaryResultsRecorder.close();
+        Path binaryResultsDir = binaryResultsRecorder.getBinaryResultsDirectory();
+
+        boolean rootTestFailed = failureMessage != null;
+
+        // Notify aggregate listener of final results
+        executionResultsListener.executionResultsAvailable(testDescriptor, binaryResultsDir, rootTestFailed);
+
+        // Throw an exception with rendered test results, if necessary
+        if (rootTestFailed) {
+            Path reportIndexFile = htmlTestReportGenerator.generateHtmlReport(testReportDirectory, binaryResultsDir);
```

Comment:
I think we should always generate the HTML report, even if there are no failures.  This is what Test does